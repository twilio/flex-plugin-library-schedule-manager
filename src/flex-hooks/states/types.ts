import * as Flex from '@twilio/flex-ui';
import { Action as ReduxAction, combineReducers } from "redux";

// Actions
export const prefix = 'CallbackAndVoicemail';
export const INITIATE_CALLBACK = `${prefix}/INITIATE_CALLBACK`;
export const REQUEUE_CALLBACK = `${prefix}/REQUEUE_CALLBACK`;
export const PLACED_CALLBACK = `${prefix}/PLACED_CALLBACK`;

// State
export interface CallbackAndVoicemailState {
	isCompletingCallbackAction: { [taskSid: string]: boolean; };
	isRequeueingCallbackAction: { [taskSid: string]: boolean; };
	lastPlacedReservationSid?: string;
};

// Extend this payload to be of type that your ReduxAction is
// Normally you'd follow this pattern...https://redux.js.org/recipes/usage-with-typescript#a-practical-example
// But that breaks the typing when adding the reducer to Flex, so no payload intellisense for you!
export interface Action extends ReduxAction {
    payload?: any;
  }
  
export const reduxNamespace = 'callbackandvoicemail';
// Register all component states under the namespace
export interface AppState {
    flex: Flex.AppState;
    [reduxNamespace]: CallbackAndVoicemailState;
}
