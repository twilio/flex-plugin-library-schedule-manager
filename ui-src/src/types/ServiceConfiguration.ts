import * as Flex from '@twilio/flex-ui';

export default interface CallbackAndVoicemailConfig {
  enabled: boolean;
  allow_requeue: boolean;
  max_attempts: number;
  auto_select_task: boolean;
}

type FlexUIAttributes = Flex.ServiceConfiguration['ui_attributes'];

export interface UIAttributes extends FlexUIAttributes {
  custom_data: {
    serverless_functions_domain: string;
    serverless_functions_protocol: string;
    serverless_functions_port: string;
    features: {
      callbacks: CallbackAndVoicemailConfig;
    };
    schedule_manager: {
      serverless_functions_domain: string;
    };
  };
}
