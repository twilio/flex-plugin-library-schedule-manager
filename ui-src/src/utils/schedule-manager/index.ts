import { Manager, Notifications } from '@twilio/flex-ui';
import { NotificationIds } from '../../flex-hooks/notifications/ScheduleManager';
import ScheduleManagerService from '../serverless/ScheduleManager/ScheduleManagerService';
import { Rule, Schedule, ScheduleManagerConfig } from '../../types/schedule-manager';
import { ErrorManager, FlexPluginErrorType } from '../../ErrorManager';
import { RRule } from 'rrule';
import { DateTime } from 'luxon';

let config = {
  data: {
    rules: [],
    schedules: [],
  },
  version: '',
} as ScheduleManagerConfig;

const delay = async (ms: number): Promise<void> => {
  return await new Promise((resolve) => setTimeout(resolve, ms));
};

export const canShowScheduleManager = (manager: Manager) => {
  const { roles } = manager.user;
  return roles.indexOf('admin') >= 0;
};

export const loadScheduleData = async (): Promise<ScheduleManagerConfig | null> => {
  const listSchedulesResponse = await ScheduleManagerService.list();

  if (listSchedulesResponse) {
    config = listSchedulesResponse;
  }

  return listSchedulesResponse;
};

export const referencedSchedules = (schedules: Schedule[], rule: Rule): string[] => {
  const refSchedules = [] as string[];
  if (schedules) {
    schedules.forEach((schedule) => {
      if (rule && schedule.rules.indexOf(rule.id) >= 0) {
        refSchedules.push(schedule.name);
      }
    });
  }

  return refSchedules;
};
export const updateScheduleData = (newSchedule: Schedule | null, existingSchedule: Schedule | null): Schedule[] => {
  if (existingSchedule === null && newSchedule !== null) {
    // adding schedule
    console.log('config.data.schedules', config.data.schedules);
    config.data.schedules = [...config.data.schedules, newSchedule];
  } else if (existingSchedule !== null && newSchedule === null) {
    // removing existing schedule
    const existingIndex = config.data.schedules.indexOf(existingSchedule);

    if (existingIndex >= 0) {
      config.data.schedules[existingIndex].isDeleted = true;
      // config.data.schedules.splice(existingIndex, 1);
    }
  } else if (existingSchedule !== null && newSchedule !== null) {
    // updating existing schedule
    const existingIndex = config.data.schedules.indexOf(existingSchedule);

    if (existingIndex >= 0) {
      config.data.schedules.splice(existingIndex, 1, newSchedule);
    }
  }

  return config.data.schedules;
};

export const updateRuleData = (newRule: Rule | null, existingRule: Rule | null): Rule[] => {
  if (existingRule === null && newRule !== null) {
    // adding rule
    config.data.rules = [...config.data.rules, newRule];
  } else if (existingRule !== null && newRule === null) {
    // removing existing rule
    const existingIndex = config.data.rules.indexOf(existingRule);

    if (existingIndex >= 0) {
      config.data.rules[existingIndex].isDeleted = true;
      // config.data.rules.splice(existingIndex, 1);
    }
  } else if (existingRule !== null && newRule !== null) {
    // updating existing rule
    const existingIndex = config.data.rules.indexOf(existingRule);

    if (existingIndex >= 0) {
      config.data.rules.splice(existingIndex, 1, newRule);
    }
  }

  return config.data.rules;
};

export const isScheduleUnique = (newSchedule: Schedule, existingSchedule: Schedule | null): boolean => {
  if (existingSchedule !== null) {
    const otherSchedules = config.data.schedules.filter((item) => existingSchedule.name !== item.name);
    const matchingSchedules = otherSchedules.filter((item) => newSchedule.name === item.name);
    return matchingSchedules.length == 0;
  } else {
    const matchingSchedules = config.data.schedules.filter((item) => newSchedule.name === item.name);
    return matchingSchedules.length == 0;
  }
};

export const isRuleUnique = (newRule: Rule, existingRule: Rule | null): boolean => {
  if (existingRule !== null) {
    const otherRules = config.data.rules.filter((item) => existingRule.id !== item.id);
    const matchingRules = otherRules.filter((item) => newRule.name === item.name);
    return matchingRules.length == 0;
  } else {
    const matchingRules = config.data.rules.filter((item) => newRule.name === item.name);
    return matchingRules.length == 0;
  }
};

export const publishSchedules = async (): Promise<number> => {
  // return values: 0=success, 2=version error, 3=failure, 4=in available activity
  if (Manager.getInstance().store.getState().flex.worker.activity.available === true) {
    Notifications.showNotification(NotificationIds.PUBLISH_FAILED_ACTIVITY);
    return 4;
  }

  const latestData: ScheduleManagerConfig = {
    data: {
      rules: config.data.rules.filter((rule) => rule.isDeleted !== true),
      schedules: config.data.schedules.filter((schedule) => schedule.isDeleted !== true),
    },
    version: config.version,
    versionIsDeployed: config.versionIsDeployed,
  };
  latestData.data.rules.map((rule) => {
    rule.isPublished && delete rule.isPublished;
    rule.isDeleted && delete rule.isDeleted;
  });
  latestData.data.schedules.map((schedule) => {
    schedule.isPublished && delete schedule.isPublished;
    schedule.isDeleted && delete schedule.isDeleted;
  });

  const updateResponse = await ScheduleManagerService.update(latestData);

  if (!updateResponse.success) {
    console.log('Schedule update failed', updateResponse);
    ErrorManager.createAndProcessError('Schedule update failed', {
      type: FlexPluginErrorType.action,
      context: 'Plugin.ScheduleManager.publishSchedules',
    });

    if (updateResponse.buildSid == 'versionError') {
      Notifications.showNotification(NotificationIds.PUBLISH_FAILED_OTHER_UPDATE);
      return 2;
    }

    Notifications.showNotification(NotificationIds.PUBLISH_FAILED);
    return 3;
  }

  // the build will take several seconds. use delay and check in a loop.
  await delay(2000);
  let updateStatus = await ScheduleManagerService.updateStatus(updateResponse.buildSid);

  while (updateStatus.buildStatus !== 'completed') {
    if (updateStatus.buildStatus === 'failed' || updateStatus.buildStatus === 'error') {
      // oh no
      console.log('Schedule update build failed', updateStatus);
      ErrorManager.createAndProcessError('Schedule update build failed', {
        type: FlexPluginErrorType.action,
        context: 'Plugin.ScheduleManager.publishSchedules',
      });
      Notifications.showNotification(NotificationIds.PUBLISH_FAILED);
      return 3;
    }

    await delay(2000);
    updateStatus = await ScheduleManagerService.updateStatus(updateResponse.buildSid);
  }

  const publishResponse = await ScheduleManagerService.publish(updateResponse.buildSid);

  if (!publishResponse.success) {
    console.log('Schedule publish failed', publishResponse);
    ErrorManager.createAndProcessError('Schedule publish failed', {
      type: FlexPluginErrorType.action,
      context: 'Plugin.ScheduleManager.publishSchedules',
    });
    Notifications.showNotification(NotificationIds.PUBLISH_FAILED);
    return 3;
  }

  Notifications.showNotification(NotificationIds.PUBLISH_SUCCESS);
  return 0;
};

const checkDate = (rule: Rule, now: Date) => {
  // first, check the date component of the rule.

  // if bounds are provided, check current date against them
  if (rule.startDate) {
    const startDate = new Date(Date.parse(rule.startDate));

    if (now.valueOf() < startDate.valueOf()) {
      // before start date; this rule is not a match
      console.log(`Rule ${rule.name} doesn't match: before start date`, startDate);
      return false;
    }
  }

  if (rule.endDate) {
    const endDate = new Date(Date.parse(rule.endDate));

    if (now.valueOf() > endDate.valueOf()) {
      // after end date; this rule is not a match
      console.log(`Rule ${rule.name} doesn't match: after end date`, endDate);
      return false;
    }
  }

  if (rule.dateRRule) {
    // extract rule from settings
    const ruleOptions = RRule.parseString(rule.dateRRule);

    // since the rule concerns only dates, and we are only comparing to today, we only need one occurrence
    // by default, past occurrences are not returned
    ruleOptions.count = 1;
    ruleOptions.dtstart = now;

    const rrule = new RRule(ruleOptions);
    let matchFound = false;

    rrule.all().forEach((occurrence) => {
      if (now.valueOf() == occurrence.valueOf()) {
        console.log(`Rule ${rule.name} occurrence matched`, occurrence);
        matchFound = true;
      }
    });

    if (!matchFound) {
      console.log(`Rule ${rule.name} doesn't match: no recurrence match`, ruleOptions);
      return false;
    }
  }

  return true;
};

const checkTime = (rule: Rule, nowTz: DateTime) => {
  if (rule.startTime) {
    const startTime = DateTime.fromISO(rule.startTime);

    if (nowTz.hour < startTime.hour) {
      console.log(`Rule ${rule.name} doesn't match: before start hour`, startTime.hour);
      return false;
    }

    if (nowTz.hour == startTime.hour && nowTz.minute < startTime.minute) {
      console.log(`Rule ${rule.name} doesn't match: before start minute`, startTime.minute);
      return false;
    }
  }

  if (rule.endTime) {
    const endTime = DateTime.fromISO(rule.endTime);

    if (nowTz.hour > endTime.hour) {
      console.log(`Rule ${rule.name} doesn't match: after end hour`, endTime.hour);
      return false;
    }

    if (nowTz.hour == endTime.hour && nowTz.minute >= endTime.minute) {
      console.log(`Rule ${rule.name} doesn't match: after end minute`, endTime.minute);
      return false;
    }
  }

  return true;
};

export const evaluateSchedule = (scheduleData: Schedule, rulesData: Rule[]) => {
  const matchingRules: Rule[] = [];
  const returnData = {
    isOpen: false,
    closedReason: 'closed',
  };

  // if the schedule is manually closed, we can return immediately
  if (scheduleData.manualClose) {
    returnData.isOpen = false;
    returnData.closedReason = 'manual';
    return returnData;
  }

  // get the current date/time in the schedule's timezone
  const nowTz: DateTime = DateTime.utc().setZone(scheduleData.timeZone);

  const now = new Date(Date.UTC(nowTz.year, nowTz.month - 1, nowTz.day));

  // evaluate each schedule rule to see if it is currently matching
  scheduleData.rules.forEach((ruleId: string) => {
    const rule = rulesData.find((rule: Rule) => rule.id == ruleId);
    if (rule && checkDate(rule, now) && checkTime(rule, nowTz)) {
      matchingRules.push(rule);
    }
  });

  const closedMatches = matchingRules.filter((rule) => !rule.isOpen);

  // if any non-open schedule(s) match, we are closed
  if (closedMatches.length > 0) {
    returnData.isOpen = false;
    returnData.closedReason = closedMatches[0].closedReason;
  } else {
    const openMatches = matchingRules.filter((rule) => rule.isOpen);

    // if only open schedule(s) match, we are open
    if (openMatches.length > 0) {
      returnData.isOpen = true;
      returnData.closedReason = '';
    }
  }

  // if nothing matches, the default returnData is used

  return returnData;
  // }
};
