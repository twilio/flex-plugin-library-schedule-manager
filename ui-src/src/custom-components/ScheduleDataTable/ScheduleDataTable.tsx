import React, { useEffect, useState } from 'react';
import { ColumnDefinition, DataTable } from '@twilio/flex-ui';
import { Button } from '@twilio-paste/core/button';
import { Box } from '@twilio-paste/core/box';
import { PlusIcon } from '@twilio-paste/icons/cjs/PlusIcon';

import ScheduleEditor from '../ScheduleEditor/ScheduleEditor';
import { Rule, Schedule } from '../../types/schedule-manager';
import ScheduleManagerStrings, { StringTemplates } from '../../flex-hooks/strings/ScheduleManager';
import { TableContainer } from '../ScheduleView/ScheduleViewStyles';
import { Badge, Flex, Heading, Popover, PopoverButton, PopoverContainer, Separator, Tooltip } from '@twilio-paste/core';
import { StatusBadge } from '@twilio-paste/status';
import SchedulesAction from '../common/SchedulesAction';
import AlertBox from '../common/AlertBox';
import { updateScheduleData } from '../../utils/schedule-manager';
import { analytics, Event} from '../../utils/Analytics';

interface OwnProps {
  isLoading: boolean;
  rules: Rule[];
  schedules: Schedule[];
  updateSchedules: (schedules: Schedule[]) => void;
  updated: Date;
}

const ScheduleDataTable = (props: OwnProps) => {
  const maxRulesLength = 60;

  const [showPanel, setShowPanel] = useState(false);
  const [selectedSchedule, setSelectedSchedule] = useState(null as Schedule | null);
  const [statusTimestamp, setStatusTimestamp] = useState('');
  // const [openIndexNext, setOpenIndexNext] = useState(null as number | null);
  const [deleteSchedule, setDeleteSchedule] = useState(null as Schedule | null);
  const [copySchedule, setCopySchedule] = useState(false);

  useEffect(() => {
    setStatusTimestamp(`${props.updated.toLocaleTimeString()} (${Intl.DateTimeFormat().resolvedOptions().timeZone})`);
  }, [props.updated]);

  useEffect(() => {
    if (selectedSchedule !== null) {
      setShowPanel(true);
    }
  }, [selectedSchedule]);

  const createScheduleClick = () => {
    analytics.track(Event.OPHRS_CREATE_SCHEDULE);
    setSelectedSchedule(null);
    setShowPanel(true);
  };

  const onPanelClosed = () => {
    setShowPanel(false);
    setSelectedSchedule(null);
  };

  const onRowClick = (item: Schedule) => {
    setSelectedSchedule(item);
  };

  const getPublishedStatus = (newSchedule: Schedule) => {
    if (newSchedule.isPublished === false || newSchedule.isDeleted === true) {
      return (
        <StatusBadge as="span" variant="ProcessWarning">
          Not published
        </StatusBadge>
      );
    } else {
      return (
        <StatusBadge as="span" variant="ProcessSuccess">
          Published
        </StatusBadge>
      );
    }
  };

  const onUpdateSchedule = (newSchedules: Schedule[]) => {
    props.updateSchedules(newSchedules);
    document.querySelector('#schedule-data-table-root')?.scrollIntoView({ behavior: 'smooth' });

    setShowPanel(false);
    setSelectedSchedule(null);
    setCopySchedule(false);
  };

  const onEditOrClone = (item: Schedule, type: string) => {
    if (type == 'copy') {
      analytics.track(Event.OPHRS_DUPLICATE_SCHEDULE,{
        scheduleName: item,
      });
      setCopySchedule(true);
    }
    if (type == 'edit') {
      analytics.track(Event.OPHRS_EDIT_SCHEDULE, {
        scheduleName: item,
      });
    }
    setSelectedSchedule(item);
  };

  const getScheduleStatus = (schedule: Schedule): any => {
    if (schedule.status) {
      const { isOpen, closedReason } = schedule.status;

      if (schedule.manualClose) {
        return (
          <Tooltip text={closedReason}>
            <Badge as="span" variant="warning">
              {ScheduleManagerStrings[StringTemplates.MANUALLYCLOSE]}
            </Badge>
          </Tooltip>
        );
      }
      if (isOpen) {
        return (
          <Badge as="span" variant="success">
            {ScheduleManagerStrings[StringTemplates.OPEN]}
          </Badge>
        );
      }

      if (closedReason.toLowerCase() === 'closed') {
        return (
          <Badge as="span" variant="warning">
            {ScheduleManagerStrings[StringTemplates.CLOSED]}
          </Badge>
        );
      } else {
        return (
          <Badge as="span" variant="warning">
            {`${ScheduleManagerStrings[StringTemplates.CLOSED]} (${closedReason})`}
          </Badge>
        );
      }
    }
  };

  const getScheduleRules = (schedule: Schedule): string => {
    const ruleNames = [] as string[];

    schedule.rules.forEach((ruleGuid) => {
      const matchingRule = props.rules.find((rule) => rule.id == ruleGuid);

      if (matchingRule) {
        ruleNames.push(matchingRule.name);
      }
    });

    return ruleNames.join(', ');
  };

  const handleDeleteConfirmSubmit = (): void => {
    if (!deleteSchedule) {
      return;
    }

    deleteSchedule.isDeleted = true;
    const newScheduleData = updateScheduleData(null, deleteSchedule);
    onUpdateSchedule([...newScheduleData]);
    setDeleteSchedule(null);
    setSelectedSchedule(null);
  };

  return (
    <>
      <div id="schedule-data-table-root">
        <Box padding="space60" display={'flex'} justifyContent={'flex-end'}>
          <Button variant="primary" disabled={props.isLoading} onClick={createScheduleClick}>
            <PlusIcon decorative />
            {ScheduleManagerStrings[StringTemplates.CREATE_SCHEDULE_BUTTON]}
          </Button>
        </Box>
        <TableContainer>
          <DataTable items={props.schedules} isLoading={props.isLoading} defaultSortColumn="name-column">
            <ColumnDefinition
              key="name-column"
              header={ScheduleManagerStrings[StringTemplates.NAME]}
              sortDirection="asc"
              sortingFn={(a: Schedule, b: Schedule) => (a.name > b.name ? 1 : -1)}
              content={(item: Schedule) => <span>{item.name}</span>}
            />
            <ColumnDefinition
              key="status-column"
              header={ScheduleManagerStrings[StringTemplates.COLUMN_STATUS]}
              subHeader={
                props.isLoading
                  ? ''
                  : `${ScheduleManagerStrings[StringTemplates.COLUMN_STATUS_ASOF]} ${statusTimestamp}`
              }
              sortingFn={(a: Schedule, b: Schedule) => (getScheduleStatus(a) > getScheduleStatus(b) ? 1 : -1)}
              content={(item: Schedule) => getScheduleStatus(item)}
            />
            <ColumnDefinition
              key="rules-column"
              header={ScheduleManagerStrings[StringTemplates.RULES]}
              sortingFn={(a: Schedule, b: Schedule) => (getScheduleRules(a) > getScheduleRules(b) ? 1 : -1)}
              content={(item: Schedule) => {
                const ruleStr = getScheduleRules(item);
                let trimmed = ruleStr;

                if (trimmed.length > maxRulesLength) {
                  trimmed = trimmed.slice(0, maxRulesLength) + '...';
                  return (
                    <PopoverContainer baseId="schedule-rules">
                      <PopoverButton variant="reset">{trimmed}</PopoverButton>
                      <Popover aria-label="Rules Popover">
                        <Box as="h3" fontWeight={'fontWeightSemibold'} fontSize={'fontSize50'} marginBottom={'space30'}>
                          {ScheduleManagerStrings[StringTemplates.RULES_DETAILS_TITLE]}
                        </Box>
                        <Box as="span">{ruleStr}</Box>
                      </Popover>
                    </PopoverContainer>
                  );
                }

                return <span>{trimmed}</span>;
              }}
            />
            <ColumnDefinition
              key="timezone-column"
              header={ScheduleManagerStrings[StringTemplates.TIMEZONE]}
              sortingFn={(a: Schedule, b: Schedule) => (a.timeZone > b.timeZone ? 1 : -1)}
              content={(item: Schedule) => <span>{item.timeZone}</span>}
            />
            <ColumnDefinition
              key="publish-status-column"
              header={ScheduleManagerStrings[StringTemplates.COLUMN_PUBLISHSTATUS]}
              // content={(item: Schedule) => {
              //   return (
              //     <StatusBadge as="span" variant="ProcessSuccess">
              //       Published
              //     </StatusBadge>
              //   );
              // }}
              content={(item: Schedule) => getPublishedStatus(item)}
            />
            <ColumnDefinition
              key="actions-column"
              header=""
              content={(item: Schedule) => (
                <SchedulesAction
                  onCopy={() => {
                    onEditOrClone(item, 'copy');
                  }}
                  deleteDisabled={false}
                  editDisabled={false}
                  copyDisabled={false}
                  onDelete={() => {
                    analytics.track(Event.OPHRS_DELETE_SCHEDULE, {
                      scheduleName: item,
                    });
                    setDeleteSchedule(item);
                  }}
                  onEdit={() => onEditOrClone(item, 'edit')}
                />
              )}
            />
          </DataTable>
        </TableContainer>
      </div>
      <ScheduleEditor
        onPanelClosed={onPanelClosed}
        rules={props.rules}
        showPanel={showPanel}
        onDelete={() => {
          analytics.track(Event.OPHRS_DELETE_SCHEDULE, {
            scheduleName: selectedSchedule,
          });
          setDeleteSchedule(selectedSchedule);
        }}
        copy={copySchedule}
        selectedSchedule={selectedSchedule}
        onUpdateSchedule={onUpdateSchedule}
      />
      <AlertBox
        handleClose={() => setDeleteSchedule(null)}
        isOpen={!!deleteSchedule}
        handleSubmit={() => handleDeleteConfirmSubmit()}
        title={ScheduleManagerStrings[StringTemplates.DELETE_CONFIRM_SCHEDULE_TITLE]}
      >
        <Flex>
          <Box as="p">{ScheduleManagerStrings[StringTemplates.DELETE_CONFIRM_SCHEDULE]}</Box>
          <Box as="strong" fontWeight={'fontWeightSemibold'} marginLeft={'space10'}>
            {deleteSchedule?.name}
          </Box>
        </Flex>
      </AlertBox>
    </>
  );
};

export default ScheduleDataTable;
