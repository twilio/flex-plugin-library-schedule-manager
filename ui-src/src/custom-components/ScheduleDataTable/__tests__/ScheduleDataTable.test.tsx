import React from 'react';
import renderer from 'react-test-renderer';
import { Schedule } from '../../../types/schedule-manager';
import ScheduleDataTable from '../ScheduleDataTable';

const renderTree = (tree) => renderer.create(tree);
describe('<ScheduleDataTable>', () => {
  it('should render component', () => {
    expect(
      renderTree(
        <ScheduleDataTable
          isLoading={false}
          rules={[]}
          schedules={[]}
          updateSchedules={(schedules: Schedule[]) => {}}
          updated={new Date()}
        />,
      ).toJSON(),
    ).toMatchSnapshot();
  });
});
