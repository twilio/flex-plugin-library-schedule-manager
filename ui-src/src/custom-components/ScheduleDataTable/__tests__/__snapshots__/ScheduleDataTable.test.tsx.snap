// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScheduleDataTable> should render component 1`] = `
<div
  id="schedule-data-table-root"
>
  <div
    className="css-cuork0"
    data-paste-element="BOX"
  >
    <button
      aria-busy="false"
      className="css-175kaqp"
      data-paste-element="BUTTON"
      disabled={false}
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      type="button"
    >
      <span
        className="css-z6jp9v"
        data-paste-element="BOX"
      >
        <span
          className="css-144s81u"
          data-paste-element="ICON"
          size="sizeIcon30"
        >
          <svg
            aria-hidden={true}
            aria-labelledby="PlusIcon-1"
            height="100%"
            role="img"
            viewBox="0 0 20 20"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.043 9.043h-4.086V4.957a.958.958 0 00-1.914 0v4.086H4.957a.958.958 0 000 1.914h4.086v4.086a.958.958 0 001.914 0v-4.086h4.086a.958.958 0 000-1.914z"
              fill="currentColor"
              fillRule="evenodd"
            />
          </svg>
        </span>
        Create new schedule
      </span>
    </button>
  </div>
  <div
    className="css-1fha43i"
  />
</div>
`;
